core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/OAbstractPaginatedStorage.java

```diff
@@ -1912,6 +1917,13 @@ public void preallocateRids(final OTransactionInternal clientTx) {
       final List<ORecordOperation> result = new ArrayList<>();
       stateLock.acquireReadLock();
       try {
+        if (pessimisticLock) {
+          for (ORecordOperation recordOperation : recordOperations) {
+            if (recordOperation.type == ORecordOperation.UPDATED || recordOperation.type == ORecordOperation.DELETED) {
+              acquireWriteLock(recordOperation.getRID());
```

Comment:
@tglman I did not find that you sort locked records by rids before locking, but you use exclusive locks not read locks. Are they already sorted or you just missed that step? Could you clarify?